HACKING.md: update the CMake related remarks

tchaikov opened the pull request:
the rules generated with CMake script is feature-wise on par with the one generated directly with `configure.py`, so the existing document is now stale.

in this change, we update the document to reflect the current status of CMake support.

---

this is an improvement in the document targeting developers, hence no need to backport.

lersek commented:
@tchaikov I think we need a larger documentation surgery for this.

The section you are currently modifying is under heading `Example: development on Fedora 25`. That "example" chapter seems irrelevant for the significant "generate ninja build files with cmake" feature. I suggest removing the current cmake references in `HACKING.md` altogether, and writing a more comprehensive / substantial description elsewhere.

1. The short example [here](https://github.com/scylladb/scylladb/blob/master/README.md#building-scylla) might want to name cmake.
2. This [document](https://github.com/scylladb/scylladb/blob/master/docs/dev/building.md) knows nothing about cmake, but it will not work when configuration occurred with cmake. For example, it contains a command `ninja <mode>-build`, which will not work when using cmake for configuration.
3. In [this section](https://github.com/scylladb/scylladb/blob/master/HACKING.md#build-system) there are again commands like `ninja-build release`, which will not work with cmake (to my understanding).

I'd like to see documentation that advertises the following command (with each part unpacked and explained):

~~~
ninja -C build {scylla,tests}:{Debug,Dev,RelWithDebInfo}
~~~

Currently, none of the above-mentioned three documents (`README.md`, `HACKING.md`, `docs/dev/building.md`) contain `RelWithDebInfo`, for example. 

In fact, if I grep the entire scylla tree for `--use-cmake`, the only hit is in `configure.py` itself; there is zero documentation on it.

A new developer should be able to read a single `.md` file and understand the complete CMake-based workflow:
- configuration,
- building the scylla server in all, or in some, configs,
- building the entire unit test suite, or some of the tests, in all, or in some, configs,
- running the test suite (all of it, or just some of it).

We have lots of stale build documentation scattered all over the place, and little  up-to-date documentation. I'd rather see all of the old build docs being deleted, and a *single* new document explaining { dbuild, native build } x { with cmake, without cmake }.

tchaikov commented:
@lersek hi László, thank you for the detailed review. before making the CMake-based building system the formal one, we need to provide a more comprehensive document to cover for instance the `<target>:<config>` notation. i will try to address *some* of your comments. as i don't intend to maintain both building system at the same time in the long run, so i think the first steps would be

1. add the CMake based document so that it is at least be able to provide enough support to a new developer to build scylla and its tests. in which, "running the test suite (all of it, or just some of it)." should have been taken care of by `test.py`. so unless you think the current document does not apply to `configure.py` or is not sufficient, i will skip this one.
2. fix #2717 , so that our CI starts using CMake for building scylla and tests.
3. revisit the related documents and restructure them if necessary

please note, i am for this proposal:

> removing the current cmake references in HACKING.md altogether, and writing a more comprehensive / substantial description elsewhere.

but i wanted to start with something simple and updated. and then expand it to a more comprehensive document.

lersek commented:
@tchaikov 

> as i don't intend to maintain both building system at the same time in the long run, i'd prefer removing the "without cmake" part from the my change. but i won't remove the document for the "without cmake" building process, and i will fix it when it's obvious wrong or outdated.

Sounds good. We can drop that documentation when CMake has altogether obsoleted the old `build.ninja` generator system.

> i think the first steps would be
> 
>     1. add the CMake based document so that it is at least be able to provide enough support to a new developer to build scylla and its tests with CMake.

This sounds good as well. And I do agree that, if your current pull request is the "first step of the first step", then it's fine (i.e., it should be OK to first clean up the obviously stale statements on CMake, even if we don't just yet have the complete CMake write-up in place). I guess I was concerned that this small docs update was going to be the *only* one, for CMake's sake. :)

> in which, "running the test suite (all of it, or just some of it)." should have been taken care of by `test.py`. so unless you think the current document does not apply to `configure.py` or is not sufficient, i will skip this one.

I agree that [this document](https://github.com/scylladb/scylladb/blob/master/docs/dev/testing.md) covers running tests. However, it does not cover building individual tests (nor do the other three documents I referenced before); I believe. From our previous discussions, there is for example a difference between rebuilding a single test binary:

~~~
# without cmake:
ninja build/{dev,debug,release}/test/boost/sstable_test

# with cmake (full syntax):
ninja -C build test/boost/{Dev,Debug,RelWithDebInfo}/sstable_test

# with cmake (shorthand; builds the binary in all modes that were not disabled at ./configure time):
ninja -C build sstable_test
~~~

I think we should capture this somewhere; it can save a lot of time, for getting a single test binary to pass.

> 
>     2. fix [Migrate to the CMake build system #2717](https://github.com/scylladb/scylladb/issues/2717) , so that our CI starts using CMake for building scylla and tests.
> 
>     3. revisit the related documents and restructure them if necessary
> 
> 
> please note, i am for this proposal:
> 
> > removing the current cmake references in HACKING.md altogether, and writing a more comprehensive / substantial description elsewhere.
> 
> but i wanted to start with something simple and updated. and then expand it to a more comprehensive document.

Sounds great. Thanks!

lersek commented:
@tchaikov do you want me to review this version? (with it being a draft)

tchaikov commented:
@lersek just wanted to update you: i am still working on this. it turns out to be messy if we keep two copies of the almost-same instructions side-by-side. so i think probably the better way is to wait until i #20261 gets merged and then i make cmake the default in `configure.py`.

tchaikov commented:
> @tchaikov do you want me to review this version? (with it being a draft)

no, not until i make cmake the default. by then, the document will reflect the status quo. now it is a bit ahead of time

tchaikov commented:
@lersek ping. the change migrating our CI to CMake has been merged.

lersek commented on HACKING.md:
```diff
@@ -280,14 +280,13 @@ Once the patch set is ready to be reviewed, push the branch to the public remote
 
-Scylla includes a [CMake](https://cmake.org/) file, `CMakeLists.txt`, for use only with development environments (not for building) so that they can properly analyze the source code.
+Scylla uses [CMake](https://cmake.org/) and ninja as its building system. This allows for seamless
+integration with any IDE or tool that supports CMake projects or can utilize `compile_commands.json`.
```
Hmmm, distcc may already be documented sufficiently: said part of this document says to set
```
export CCACHE_PREFIX="distcc"
```
and that way ccache takes care of distcc.

Based on that, it might suffice just to recommend

```
export CMAKE_CXX_COMPILER_LAUNCHER=/usr/bin/ccache
export CMAKE_C_COMPILER_LAUNCHER=/usr/bin/ccache
```

lersek replied:
- do we need to extend install-dependencies.sh to install cmake?
- do we need to update these documents (wherever they speak about dependencies / install-dependencies) to also mention cmake?

lersek replied:
- going forward, are we going to include cmake in the frozen toolchain (dbuild) container images? (well this question applies more to the README, not HACKING, sorry)

tchaikov replied:
> * do we need to extend install-dependencies.sh to install cmake?

it's already extended. see https://github.com/scylladb/scylladb/blob/298cafff3537950f1b17b1fcf119e3d831bcd786/install-dependencies.sh#L304

and

- https://github.com/scylladb/seastar/blob/47c52cb3e19c4757093b812876ea245d7fc0b62d/install-dependencies.sh#L31
- https://github.com/scylladb/seastar/blob/47c52cb3e19c4757093b812876ea245d7fc0b62d/install-dependencies.sh#L82
- https://github.com/scylladb/seastar/blob/47c52cb3e19c4757093b812876ea245d7fc0b62d/install-dependencies.sh#L170

> 
>     * do we need to update these documents (wherever they speak about dependencies / install-dependencies) to also mention cmake?

now that it's already covered by seastar's `install-dependencies.sh`. i think CMake is part of the building dependencies, unless we are picky on cmake versions or configurations. i don't think we should single it out, and talk about it specifically.

lersek commented on README.md:
```diff
@@ -32,7 +32,7 @@ Building Scylla with the frozen toolchain `dbuild` is as easy as:
 $ git submodule update --init --force --recursive
 $ ./tools/toolchain/dbuild ./configure.py
-$ ./tools/toolchain/dbuild ninja build/release/scylla
+$ ./tools/toolchain/dbuild ninja -C build scylla:RelWithDebInfo
```
Please confirm this actually works with cmake now (i.e., cmake is included in the container image).

lersek replied:
... the syntax looks OK to me, otherwise.

lersek replied:
totally tangential side comment: lower down in this file, we have a mention of `test.py`, with a pointer to `docs/dev/testing.md`. That document seems to say nothing about `--no-gather-metrics`, or `systemd-run` :( so I think the currently documented way will encounter cgroups permission errors. Oh well.

tchaikov replied:
> Please confirm this actually works with cmake now (i.e., cmake is included in the container image).

we run `install-dependencies.sh` when building the toolchain image, see https://github.com/scylladb/scylladb/blob/298cafff3537950f1b17b1fcf119e3d831bcd786/tools/toolchain/Dockerfile#L11 . so it should work.

tchaikov replied:
> totally tangential side comment: lower down in this file, we have a mention of `test.py`, with a pointer to `docs/dev/testing.md`. That document seems to say nothing about `--no-gather-metrics`, or `systemd-run` :( so I think the currently documented way will encounter cgroups permission errors. Oh well.

@xtrey hello Andrei, do you think we should improve the help message of `--gather-metrics` and related documentation?

xtrey replied:
@tchaikov we can improve it. But there are no issues with metrics inside the toolchain.

lersek commented on docs/dev/building.md:
```diff
@@ -22,37 +22,45 @@ git clone https://github.com/scylladb/scylla
 git submodule update --init --force --recursive
 ./tools/toolchain/dbuild ./configure.py --mode=<mode>
-./tools/toolchain/dbuild ninja
+./tools/toolchain/dbuild cmake --build build
```
This surprises me: after `configure.py`, I just run something like
```
ninja -C build {scylla,tests}:Dev
```

and indeed your update above, to `README.md`, near the `dbuild` stuff, recommends `ninja -C build scylla:RelWithDebInfo`. So this part looks at least inconsistent (even though it may perfectly well work -- I've never tried it).

tchaikov replied:
i think, `README.md` should serve as a minimal but workable solution. `building.md` can expand the topic further by providing more options. the upside of using `cmake` to drive the build is, we don't need to use the less-documented notation of `<target>:<config>` when we need to specify the build mode / config.

lersek commented on docs/dev/building.md:
```diff
@@ -22,37 +22,45 @@ git clone https://github.com/scylladb/scylla
 git submodule update --init --force --recursive
 ./tools/toolchain/dbuild ./configure.py
-./tools/toolchain/dbuild ninja <mode>
+./tools/toolchain/dbuild cmake --build build --config <config>
```
Ditto: possibly right (I couldn't say), but not consistent with the recommendation in README.md (and with my own use thus far). Please confirm.

tchaikov replied:
i see what you mean now -- "recommendation". i intentionally avoid talking too much about CMake in "README.md" to avoid the potential contentious CMake versus ninja discussions. instead, i prefer keeping the `README.md` neutral, and leave the discussion for future. 

while I'm still learning the nuances of these technical discussions, I believe this PR is important for the CMake migration path. i suggest we merge it now to maintain momentum, and we can refine the documentation's language and scope later as we gather more community feedback.

lersek commented on docs/dev/building.md:
```diff
@@ -22,37 +22,45 @@ git clone https://github.com/scylladb/scylla
 
-After the build completes, you can find build artifacts in `build/<mode>` directory.
+Instead of using `cmake`, you can use `ninja` directly with the notation of
+`<target>:<config>`:
```
Ah, nice, this explains it then. Thanks.

lersek commented on docs/dev/building.md:
````diff
@@ -22,37 +22,45 @@ git clone https://github.com/scylladb/scylla
+
+```console
+./tools/toolchain/dbuild ninja -C build tests:<config>
+
````
Unnecessary empty line?

tchaikov replied:
fixed.

lersek commented on docs/dev/building.md:
````diff
@@ -22,37 +22,45 @@ git clone https://github.com/scylladb/scylla
 
 ```console
-./tools/toolchain/dbuild ninja test
+./tools/toolchain/dbuild ninja -C build test
````
I think this instruction is probably fine; however, the leading text -- **run** unit tests -- seems wrong. This command is supposed to *build* unit tests, right?

tchaikov replied:
both the building systems generated by `configure.py` and by `cmake` have a target named `test`, which actually runs `test.py`. so i think this command does run unit tests. to build unit tests, we have another target named `tests`.

tchaikov replied:
see https://github.com/scylladb/scylladb/blob/298cafff3537950f1b17b1fcf119e3d831bcd786/configure.py#L2051-L2052 and https://github.com/scylladb/scylladb/blob/298cafff3537950f1b17b1fcf119e3d831bcd786/test/CMakeLists.txt#L118

lersek reviewed (changes requested):
I guess the only "change" I'm requesting is to document the `CMAKE_CXX_COMPILER_LAUNCHER` env var; the rest is just asking for confirmation/information. Thanks!

tchaikov commented:
v2:

- document `CMAKE_CXX_COMPILER_LAUNCHER` and `CMAKE_C_COMPILER_LAUNCHER` en vars
- remove extraneous empty line

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12770/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12770/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12456/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4434/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1366/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12770/)
#### Build Details:
- Duration: 4 hr 26 min
- Builder: i-0942547bf01b36eac (m5d.12xlarge)

lersek reviewed (approved):
Awesome, thanks!

avikivity commented on configure.py:
```diff
@@ -753,7 +753,7 @@ def find_ninja():
 arg_parser.add_argument('--date-stamp', dest='date_stamp', type=str,
                         help='Set datestamp for SCYLLA-VERSION-GEN')
-arg_parser.add_argument('--use-cmake', action=argparse.BooleanOptionalAction, default=False, help='Whether to use CMake as the build system')
+arg_parser.add_argument('--use-cmake', action=argparse.BooleanOptionalAction, default=True, help='Whether to use CMake as the build system')
```
We need proof that cmake passes the same compile options as the traditional system.

tchaikov replied:
i compared the compilation database generated from CMake and `configure.py` using the script at https://gist.github.com/tchaikov/eb2cdb08c013506346cd0c476f5b45da. the result can be found at https://gist.github.com/tchaikov/c7d5b79b7b8dfd9144ec3a27b79c4371

tchaikov replied:
> I looked at the diff and the only significant difference is this:
> 
> ```diff
> === wasmtime_bindings.cc:
> -{'W': set(), 'I': {'build/gen'}, 'f': set(), 'D': set()}
> +{'W': {'error=unused-result'}, 'I': {'build/Dev/seastar/gen/include', 'seastar/include'}, 'f': {'stack-clash-protection'}, 'D': {'SEASTAR_SCHEDULING_GROUPS_COUNT=16', 'SEASTAR_LOGGER_TYPE_STDOUT', 'SEASTAR_API_LEVEL=7', 'SEASTAR_SSTRING', 'SEASTAR_TYPE_ERASE_MORE', 'WITH_GZFILEOP', 'SEASTAR_LOGGER_COMPILE_TIME_FMT', 'FMT_SHARED'}}
> 
> === inc.cc:
> -{'W': set(), 'I': {'build/gen'}, 'f': set(), 'D': set()}
> +{'W': {'error=unused-result'}, 'I': {'build/Dev/seastar/gen/include', 'seastar/include'}, 'f': {'stack-clash-protection'}, 'D': {'SEASTAR_SCHEDULING_GROUPS_COUNT=16', 'SEASTAR_LOGGER_TYPE_STDOUT', 'SEASTAR_API_LEVEL=7', 'SEASTAR_SSTRING', 'SEASTAR_TYPE_ERASE_MORE', 'WITH_GZFILEOP', 'SEASTAR_LOGGER_COMPILE_TIME_FMT', 'FMT_SHARED'}}
> ```
> 
> Can you explain this?
> 

sure. these two files are both rust's C++ bindings generated using cxxbridge.

`build/gen`
---

in this very case, `configure.py` puts the rust module's C++ bindings into `build/<build_mode>/gen/rust`. while CMake puts them into `build/gen/rust`. as all modes share the same generated C++ bindings. BTW, we put the generated object files into per-build-mode subdirectories. and the compare script normalize the path used by `configure.py` by removing dirname, see https://gist.github.com/tchaikov/eb2cdb08c013506346cd0c476f5b45da .

seastar flags
---

rust binding does not use Seastar headers or link against Seastar libraries, so there is no need to build with Seastar cflags. this is also why i claimed 

> The variations mostly represent cleanup opportunities in the CMake configuration

in previous comment in this thread.


> BTW which is `-` and which is `+`?

i wanted to mimic the format of `diff` when comparing two building command databases. the two building command databases are generated using `configure.py` and CMake repectively, and the command is
```bash
../compare-compile-commands.py -ppppp {cmake-,}compile_commands.json
```

so the lines started with `-` is the command line options only found in the rules generated by CMake, and the lines started with `+` are those only found in the rules generated by `configure.py`. as you noticed, `configure.py` tends to pass more cflags when building object files even when they are not used at all.

> 
> This `'I': {'build/gen'},` is present everywhere, why is that?

actually, both of them have `build/gen`. but i am too lazy to normalize them. 

- the one from CMake looks like `-I/home/kefu/dev/scylladb/master -I/home/kefu/dev/scylladb/master/build/gen` 
- the one from `configure.py` looks like `-iquote. -iquote build/dev/gen`.

denesb replied:
Ok, it indeed looks like the generated build rules are identical where it counts. I'll queue this. :crossed_fingers:

yaronkaikov replied:
This is breaking gating, I assume because we need this change together with some change in `scylla-pkg`

yaronkaikov replied:
@scylladb/scylla-maint can you please dequeue this patch? it's breaking gating

tchaikov commented:
@scylladb/scylla-maint hello maintainers, could you help review this change?

avikivity commented:
Dequeued since it broke promotion.

@tchaikov the proof it doesn't change flags should be in the cover letter, not hidden in a comment.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14020/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/838/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5575/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13588/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2458/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/240/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14020/)
#### Build Details:
- Duration: 4 hr 0 min
- Builder: spider2.cloudius-systems.com

tchaikov commented:
v3:

- noted that i've compared the build commands generated by `configure.py` and those generated by CMake building system in the commit message and in the cover letter.

tchaikov commented:
@denesb hi Botond, could you help re-review this change? i've switched the CI pipelines to CMake in https://github.com/scylladb/scylla-pkg/commit/4df8b8b54f9e8599f367ccbb4917ff881c6452a6 .

avikivity commented:
@tchaikov let's wait until the build system changes are merged from enterprise. These are so large they would invalidate the comparison results.

tchaikov commented:
> @tchaikov let's wait until the build system changes are merged from enterprise. These are so large they would invalidate the comparison results.

okay. let's put this change on hold then.

tchaikov commented:
v4:

- rebased against master HEAD
- added "Fixes #22212" tag in the commit message
- updated the cover letter with the link to the updated compare result

tchaikov commented:
@avikivity and @denesb, i

1. regenerate the compare result at https://gist.github.com/tchaikov/ebeadaaf6ea6bbcebc20feae10555cd8 with the latest master HEAD.
2. checked it. no significant issues identified.

could you take another look?

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14374/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/1085/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2715/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13843/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5832/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/487/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14374/)
#### Build Details:
- Duration: 8 hr 21 min
- Builder: i-05d8fb3b868e811e5 (m5ad.8xlarge)

denesb commented:
LGTM but I'll leave for @avikivity to decide on the merge, in case it interferes with something else.

nyh commented on README.md:
```diff
@@ -32,7 +32,7 @@ Building Scylla with the frozen toolchain `dbuild` is as easy as:
 $ git submodule update --init --force --recursive
 $ ./tools/toolchain/dbuild ./configure.py
-$ ./tools/toolchain/dbuild ninja build/release/scylla
+$ ./tools/toolchain/dbuild ninja -C build scylla:RelWithDebInfo
```
I saw that `ninja -C build Dev/scylla` works, so why write `scylla:Dev` with a colon? The benefit of `Dev/scylla` is that it's the actual path of build/Dev/scylla, which you'll probably need to know.
Also, what the heck is "RelWithDebInfo" and is this really the simplest example we can give?

tchaikov replied:
i am using `ninja -C build Dev/scylla` in the latest revision. "RelWithDebInfo" is one of the typical `CMAKE_BUILD_TYPE`. it is the closest thing to "release". see https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html

nyh commented on docs/dev/building.md:
```diff
@@ -22,37 +22,44 @@ git clone https://github.com/scylladb/scylla
 git submodule update --init --force --recursive
 ./tools/toolchain/dbuild ./configure.py --mode=<mode>
-./tools/toolchain/dbuild ninja
+./tools/toolchain/dbuild cmake --build build
```
Wait, what? Aren't you supposed to continue running "ninja" to build? Also in README.md you gave an example building with ninja, so why is this one different?
I suggest that we:
1. Think of "the best" way to build - if you think it's with cmake, not ninja, than give this example in all doc files and not just one.
2. In one (or at least one) doc files also give the "other" way, in case someone is curious.

I see other reviewers commented this too. I think they were right.

tchaikov replied:
i think it's cmake. but per your comment at https://github.com/scylladb/scylladb/pull/20188/files#r1913248685 , it seems we need to stick with `ninja`. so i am using `ninja` in all doc, and provide `cmake` in `building.md` for those who are curious.

nyh commented on docs/dev/building.md:
```diff
@@ -22,37 +22,44 @@ git clone https://github.com/scylladb/scylla
 
-After the build completes, you can find build artifacts in `build/<mode>` directory.
+Instead of using `cmake`, you can use `ninja` directly with the notation of
+`<target>:<config>`:
```
I saw that config/target works too, and makes more sense...

tchaikov replied:
i am inclined to encourage developers to use `<target>:<config>` or just go with CMake. because `config/target` does not always work. what if "target" is not the name of the built artifacts, for instance, what if i just want to build the `alternator` target defined in `alternator/CMakeLists.txt`  to see if my change compiles ?

   ```console
   $ ninja -C ~/dev/scylladb/build Debug/alternator
   ninja: Entering directory `/home/kefu/dev/scylladb/build'
   ninja: error: unknown target 'Debug/alternator'
   $ ninja -C ~/dev/scylladb/build build/alternator/Debug/libalternator.a -j6
   ninja: Entering directory `/home/kefu/dev/scylladb/build'
   ninja: error: unknown target **'build/alternator/Debug/libalternator.a'**
   $ ninja -C ~/dev/scylladb/build alternator:Debug
   ninja: Entering directory `/home/kefu/dev/scylladb/build'
   [1/14] Performing build step for 'Seastar'
   ninja: no work to do.
   [2/3] No install step for 'Seastar'
   [3/3] Completed 'Seastar'
   ```

nyh commented on docs/dev/building.md:
````diff
@@ -22,37 +22,44 @@ git clone https://github.com/scylladb/scylla
+./tools/toolchain/dbuild cmake --build build
 ```
 
 where `mode` is `dev` for development builds, `release` for release builds, and `debug` for debug builds.
````
This line is out of date and must be fixed. The names are no different.

tchaikov replied:
i think this line is not outdated. `dev` , `release` and `debug` are still accepted by `configure.py`.

nyh commented on docs/dev/building.md:
```diff
@@ -22,37 +22,44 @@ git clone https://github.com/scylladb/scylla
 where `mode` is `dev` for development builds, `release` for release builds, and `debug` for debug builds.
 
-Alternatively, you can also pass the `mode` directly to the `ninja` command:
+Alternatively, you can also pass the `config` directly to the `cmake` command:
```
For some reason one line after you explain what "mode" is, you switch to calling it "config". The name "config" is somewhat ambiguous (since just a few lines above you used the command "configure" and called it "mode", not "config", but if you want to use `<config>` here as you do you need to explain what it is, and what values it can take.

tchaikov replied:
> For some reason one line after you explain what "mode" is, you switch to calling it "config".

i am not using different name to represent "it". "mode" only applies to `configure.py`, while `config` applies to CMake and ninja.
 
>  The name "config" is somewhat ambiguous (since just a few lines above you used the command "configure" and called it "mode", not "config", but if you want to use `<config>` here as you do you need to explain what it is, and what values it can take.

i see. but i don't want to put all the details in "Getting Started". as this section should be relatively short. we actually expanded it in "Configuring" section. probably i should just use `cmake` in "Getting Started". and explain the mapping between mode and config, and corresponding ninja commands in following sections.

nyh commented on docs/dev/building.md:
```diff
@@ -77,20 +84,36 @@ If you don't specify a build mode, the `build.ninja` will contain configuration
 
 [Ninja]: https://ninja.org/
 
+In the CMake-based building system, the build modes are mapped to the configuration names:
```
In what sense do we still have the notion of "build modes" taking dev/release and so on? The directory is no longer build/dev, it is build/Dev, and so on. Why shold a user care about the old names?

tchaikov replied:
our toolings are built around "dev", "release", etc. and quite a few developers are used to "dev" and "release" notations, so i think i have to note this down.

nyh commented on docs/dev/building.md:
````diff
@@ -22,37 +22,44 @@ git clone https://github.com/scylladb/scylla
 
 ```console
-./tools/toolchain/dbuild ninja test
+./tools/toolchain/dbuild ninja -C build test
````
I think this is not enough - I almost never want to build, let alone, all tests, and want to compile a single test. So I need to know a command line like:
```
ninja -C build test/boost/Dev/sstable_datafile_test
```
I would have also liked to see an explanation of where cmake puts object files (like `test/boost/CMakeFiles/sstable_datafile_test.dir/Debug/sstable_datafile_test.cc.o` - who can guess that???) but I guess it doesn't belong in this specific place. But I think it needs to appear somewhere. Developers will need to understand the new organization of build/, which is very different from what it was.

tchaikov replied:
i think "Getting Started" should provide the minimal information to get a developer started. i will update the "Testing" section with more details on how to build a single test.

nyh commented on docs/dev/building.md:
````diff
@@ -77,20 +84,36 @@ If you don't specify a build mode, the `build.ninja` will contain configuration
+./tools/toolchain/dbuild ninja -C build
 ```
 
 This will build Scylla for the build mode specified by the `configure.py` command.
````
Please audit all the places where the words "build mode" or "config" are used, and change all of them to use the same terminology. You can mention in one place also the older or alternative terminology - but it's not good that you randomly switch between the two terminologies.

tchaikov replied:
please note, i am not using these two terminologies interchangeably, or "randomly" switching between them. "mode" is used along with `configure.py`, while `config` is used along with CMake and ninja.

tchaikov replied:
ahh, i did find some places where i was using "mode" by mistake. corrected.

nyh commented on docs/dev/building.md:
```diff
@@ -138,41 +166,41 @@ This section lists all the artifacts generated by the build system.
 **Tests**:
 
-* `build/<mode>/test/**`
+* `build/<config>/test/**`
```
This doesn't seem correct - I see for example build/test/boost/Dev/sstable_datafile_test - the `<config>` is **not** a top-level subdirectory of build. Which I am sad about, because I was used to doing "rm -r build/debug" to delete an unwanted build mode, and I can no longer do that.

tchaikov replied:
indeed. should be fixed in the latest revision.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14406/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/1113/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13878/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5867/)
:x: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/518/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2749/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14406/)
#### Build Details:
- Duration: 8 hr 14 min
- Builder: i-07134a4824efa8457 (m5ad.8xlarge)

tchaikov commented:
v5:

- moved the cmake command examples from the "Getting Started" section to "Configuring" section
- replaced "mode" with "config" in some places to reduce the confusion
- corrected the path of the test executables 
- added the layout of the built artifacts.

tchaikov commented:
> > LGTM but I'll leave for @avikivity to decide on the merge, in case it interferes with something else.
> 
> We now need an extended proof due to the introduction of pgo/lto. Sorry for making this so hard, but regressions here are very difficult to find, so let's do our best to rule them out ahead of time.

the compare script checks the compiling commands. i will improve it so it compares the link commands as well.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14479/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13936/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2805/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5925/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14479/)
#### Build Details:
- Duration: 11 hr
- Builder: i-0c991f01fff518e34 (m5ad.8xlarge)

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14683/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2956/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14090/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6079/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14683/)
#### Failed Tests (3/37276):
- [test_rolling_upgrade](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14683/testReport/junit/enterprise_upgrade_test/TestEnterpriseRollingUpgrade/Tests___dtest_with_tablets___test_rolling_upgrade) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_rolling_upgrade)
- [test_rolling_upgrade](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14683/testReport/junit/enterprise_upgrade_test/TestEnterpriseRollingUpgrade/Tests___dtest_with_topology_changes___test_rolling_upgrade) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_rolling_upgrade)
- [test_rolling_upgrade](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14683/testReport/junit/enterprise_upgrade_test/TestEnterpriseRollingUpgrade/Tests___dtest___test_rolling_upgrade) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_rolling_upgrade)
#### Build Details:
- Duration: 10 hr
- Builder: i-07925296504bf887e (m5ad.8xlarge)

mykaul commented:
@tchaikov - what's the status of this PR?

tchaikov commented:
> @tchaikov - what's the status of this PR?

@mykaul  i am supposed to work on the test plan for s3 tests at this moment. so everything else has lower priority. i was trying to capture the link flags with and without this change.

mykaul commented:
Close for the time being?

avikivity commented:
> > Close for the time being?
> 
> @mykaul @avikivity I think it's about time (even belated) to make a decision: Are we switching to cmake, or not?
> 
> 1. If we are, why are we not doing it now? We've been dragging this for several years now. What will we gain by waiting several more years?
> 2. If we're not, let's stop pretending that we're switching to cmake, drop the cmake stuff and start focusing on configure.py in earnest.
> 
> I'm personally not fond of cmake at all, but I don't feel I know it well enough to say anything concrete **against** it, and clearly there are many developers who swear by cmake, so I won't be against this move if it's done - but we need to make a decision - are we doing it, or not.

We can't switch without a proof that we haven't lost any compile flags.

nyh commented:
> We can't switch without a proof that we haven't lost any compile flags.

How do we plan to do this?

If you **want** to switch to cmake, I think the path is quite obvious: Switch to it in master now, assuming that @tchaikov's past work already solved most of the problems (and people didn't break things afterwards), and then deal with the fallout. We can do this after branching 2025.3, so if we have teething problems - broken builds, reduced performance, or who knows what - we will have plenty of time to solve all of them.

Alternatively, you can decide **not to switch** to cmake, ever. I wouldn't shed any tears if this will be decision.

But what I don't understand is the idea that "let's switch in the future, but just not now". This is exactly what we've been doing for several years, and the "future" never becomes now. In the meantime everyone is avoiding improving configure.py because we keep saying that nobody cares about it, it will soon go away, and everybody keeps sending patches which are only tested with configure.py, not cmake, so the difference between the two just grows - rather than shrink.

avikivity commented:
> > We can't switch without a proof that we haven't lost any compile flags.
> 
> How do we plan to do this?

Build a candidate translation unit, collect the command line flags, sort, break into lines, diff.

> 
> If you **want** to switch to cmake, I think the path is quite obvious: Switch to it in master now, assuming that @tchaikov's past work already solved most of the problems (and people didn't break things afterwards), and then deal with the fallout. We can do this after branching 2025.3, so if we have teething problems - broken builds, reduced performance, or who knows what - we will have plenty of time to solve all of them.

Many we won't discover. For example, if we have fewer -Wsomething, we just won't see warnings on new code.

> 
> Alternatively, you can decide **not to switch** to cmake, ever. I wouldn't shed any tears if this will be decision.
> 

I have nothing positive to say about cmake.


> But what I don't understand is the idea that "let's switch in the future, but just not now". This is exactly what we've been doing for several years, and the "future" never becomes now. In the meantime everyone is avoiding improving configure.py because we keep saying that nobody cares about it, it will soon go away, and everybody keeps sending patches which are only tested with configure.py, not cmake, so the difference between the two just grows - rather than shrink.

Nobody is assigned to it, so it doesn't happen.

mykaul commented:
It's a reasonable 20% project, if someone cares deeply about it, btw.

nyh commented:
> It's a reasonable 20% project, if someone cares deeply about it, btw.

@mykaul, repeating what I already said above, the problem is not finding a volunteer that cares deeply about cmake. We have a problem for **everyone** right now: The problem is that for several **years** now we've been in a limbo where:

1. Developers are asked not to improve the non-cmake build system before "we're switching to cmake soon".
2. Whenever a developer does something in the non-cmake build system (even add a new source file), they are asked to also do the same thing in cmake too.
3. But because hardly anyone is using cmake (how to use it isn't even documented - see https://github.com/scylladb/scylladb/issues/22212) and neither does CI, developers often break the cmake build requiring the few cmake fans to fix those problems later (e.g, today we got https://github.com/scylladb/scylladb/pull/26690 from @kreuzerkrieg).

Do we want to continue this for several more **years**? Are you happy with this state? I guess some are happy - it lets both cmake-fanboys and cmake-haters get what they want. But in my opinion, we should make a decision: either drop cmake completely or switch to it completely. If you make a decision "switch to cmake" and it will require some work, then, obviously, someone will do this work. The other option (drop cmake) doesn't require work, but I don't know if that's the option you want to choose.

mykaul commented:
> > It's a reasonable 20% project, if someone cares deeply about it, btw.
> 
> @mykaul, repeating what I already said above, the problem is not finding a volunteer that cares deeply about cmake. We have a problem for **everyone** right now: The problem is that for several **years** now we've been in a limbo where:
> 
> 1. Developers are asked not to improve the non-cmake build system before "we're switching to cmake soon".
> 2. Whenever a developer does something in the non-cmake build system (even add a new source file), they are asked to also do the same thing in cmake too.
> 3. But because hardly anyone is using cmake (how to use it isn't even documented - see [Document how to build Scylla with cmake - like CI does #22212](https://github.com/scylladb/scylladb/issues/22212)) and neither does CI, developers often break the cmake build requiring the few cmake fans to fix those problems later (e.g, today we got [cmake: fix the seastar API level #26690](https://github.com/scylladb/scylladb/pull/26690) from @kreuzerkrieg).
> 
> Do we want to continue this for several more **years**? Are you happy with this state? I guess some are happy - it lets both cmake-fanboys and cmake-haters get what they want. But in my opinion, we should make a decision: either drop cmake completely or switch to it completely. If you make a decision "switch to cmake" and it will require some work, then, obviously, someone will do this work. The other option (drop cmake) doesn't require work, but I don't know if that's the option you want to choose.

I personally don't see the reason and value of it, but I understand there's some strong motivation to do so. I therefore not happy with current flow, but I also don't see a different way to handle it. Specifically, I'm against a 'stop-switch' move. It's too painful.
